File: apps/embedding-worker/src/circuit_breaker.py
"""Minimal in-process circuit breaker (closed / open / half-open)."""

import time
from typing import Callable, TypeVar

T = TypeVar("T")


class CircuitBreakerOpenError(Exception):
    def __init__(self, name: str):
        super().__init__(f"Circuit breaker '{name}' is open")
        self.breaker_name = name


class CircuitBreaker:
    def __init__(
        self,
        name: str,
        failure_threshold: int = 5,
        reset_timeout_sec: float = 30.0,
    ):
        self.name = name
        self.failure_threshold = failure_threshold
        self.reset_timeout_sec = reset_timeout_sec

        self._state: str = "closed"  # closed | open | half-open
        self._failure_count: int = 0
        self._last_failure_time: float = 0.0

    @property
    def state(self) -> str:
        if self._state == "open":
            if time.monotonic() - self._last_failure_time >= self.reset_timeout_sec:
                self._state = "half-open"
        return self._state

    def execute(self, fn: Callable[[], T]) -> T:
        current = self.state

        if current == "open":
            raise CircuitBreakerOpenError(self.name)

        try:
            result = fn()
            self._on_success()
            return result
        except Exception:
            self._on_failure()
            raise

    def _on_success(self) -> None:
        self._failure_count = 0
        self._state = "closed"

    def _on_failure(self) -> None:
        self._failure_count += 1
        self._last_failure_time = time.monotonic()
        if self._failure_count >= self.failure_threshold:
            self._state = "open"

    def reset(self) -> None:
        self._state = "closed"
        self._failure_count = 0
        self._last_failure_time = 0.0

File: packages/observability/src/circuitBreaker.ts
export type CircuitState = "closed" | "open" | "half-open";

export interface CircuitBreakerOptions {
  /** Name for logging/metrics. */
  name: string;
  /** Number of consecutive failures before opening. Default 5. */
  failureThreshold?: number;
  /** How long to stay open before allowing a probe (ms). Default 30_000. */
  resetTimeoutMs?: number;
  /** Optional callback when state changes. */
  onStateChange?: (
    name: string,
    from: CircuitState,
    to: CircuitState
  ) => void;
}

/**
 * Error thrown when the circuit breaker is open and not accepting requests.
 */
export class CircuitBreakerOpenError extends Error {
  public readonly breakerName: string;

  constructor(name: string) {
    super(`Circuit breaker "${name}" is open`);
    this.name = "CircuitBreakerOpenError";
    this.breakerName = name;
  }
}

/**
 * In-process circuit breaker with closed/open/half-open states.
 *
 * - **Closed**: requests pass through normally; consecutive failures are counted.
 * - **Open**: requests are rejected immediately with CircuitBreakerOpenError.
 *   After `resetTimeoutMs`, transitions to half-open.
 * - **Half-open**: one probe request is allowed. Success → closed; failure → open.
 */
export class CircuitBreaker {
  private state: CircuitState = "closed";
  private failureCount = 0;
  private lastFailureTime = 0;

  private readonly _name: string;
  private readonly failureThreshold: number;
  private readonly resetTimeoutMs: number;
  private readonly onStateChange?: (
    name: string,
    from: CircuitState,
    to: CircuitState
  ) => void;

  constructor(opts: CircuitBreakerOptions) {
    this._name = opts.name;
    this.failureThreshold = opts.failureThreshold ?? 5;
    this.resetTimeoutMs = opts.resetTimeoutMs ?? 30_000;
    this.onStateChange = opts.onStateChange;
  }

  getState(): CircuitState {
    return this.state;
  }

  getName(): string {
    return this._name;
  }

  /**
   * Execute `fn` through the breaker.
   */
  async execute<T>(fn: () => Promise<T>): Promise<T> {
    if (this.state === "open") {
      if (Date.now() - this.lastFailureTime >= this.resetTimeoutMs) {
        this.transition("half-open");
      } else {
        throw new CircuitBreakerOpenError(this._name);
      }
    }

    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (err) {
      this.onFailure();
      throw err;
    }
  }

  private onSuccess(): void {
    this.failureCount = 0;
    if (this.state !== "closed") {
      this.transition("closed");
    }
  }

  private onFailure(): void {
    this.failureCount++;
    this.lastFailureTime = Date.now();

    if (this.state === "half-open") {
      this.transition("open");
    } else if (this.failureCount >= this.failureThreshold) {
      this.transition("open");
    }
  }

  private transition(to: CircuitState): void {
    const from = this.state;
    this.state = to;
    this.onStateChange?.(this._name, from, to);
  }

  /** Reset for testing. */
  reset(): void {
    this.state = "closed";
    this.failureCount = 0;
    this.lastFailureTime = 0;
  }
}
